[[ cpptest ]] Add C++ unit testing framework

livecodestephen opened the pull request:
Add a C++ testing framework based on Google Test (https://github.com/google/googletest).  This consists of a new library, `libcpptest`, containing Google Test as a submodule, which contains all the testing infrastructure.  Each library can then contain a `tests` folder containing tests.

Todo:
- [x] Factor out `main` so that the 'kernel-*' targets can be used as a dependencies of the test executable.
- [x] Fix windows build
- [x] Add `libcpptest` library containing googletest as a submodule
- [x] Add C++ test infrastructure to `engine`
- [x] Add C++ test infrastructure to `libfoundation`
- [x] Add TAP output
- [x] Add build targets to run the tests

livecodestephen committed:
[[ cpptest ]] Factor `main` through a simple wrapper

Rename the existing `main` functions to `platform_main` and add a simple
`main` which just calls `platform_main`.  This means that, for example, the
'kernel-server' can be used as a dependency of a C++ test executable with a
different `main`.

livecodestephen commented:
@livecode-vulcan try  c84bbea

livecodestephen committed:
[[ cpptest ]] Fix Windows build

Don't include the wrapper `main` on Windows, where `WinMain` will be used
anyway.

livecodestephen committed:
[[ cpptest ]] Fix android build

The android build does need the `main` wrapper either.

livecodestephen committed:
[[ cpptest ]] Adds Google Test submodule and associated build rules

livecodestephen committed:
[[ cpptest ]] Specifies windows executable should use CONSOLE subsystem

livecodestephen committed:
[[ cpptest ]] Fix compilation error on iOS

On iOS we can't compile to a console application, so just compile the tests to a
shared library.  This means they are at least compiled, even if they can't be
run.

livecodestephen committed:
[[ cpptest ]] Move `kernel-*.gypi` to `kernel-*.gyp`

This change is so that I can include `cpptest.gypi` in each of the `kernel-*.gyp`
files.

livecodestephen committed:
[[ cpptest ]] Add C++ test framework to the engine folder

livecodestephen committed:
[[ cpptest ]] Fix windows compile errors

Adding `GTEST_HAS_EXCEPTIONS=0` doesn't actually fix any error but should
probably be set as LiveCode doesn't use exceptions.

livecodestephen committed:
Fix unchecked use of placement new

livecodestephen committed:
[[ cpptest ]] Run tests as part of build. Test memory allocation failure.

livecodestephen committed:
[[ cpptest ]] Fix build problems

livecodestephen committed:
[[ cpptest ]] Add test framework to libfoundation

livecodestephen committed:
[[ cpptest ]] Add TAP formatted test logs

A TAP test log can now be specified using a `--tap=<filename>` argument on
the test runner.

livecodestephen committed:
[[ cpptest ]] Move emscripten-javascriptify.py to util

livecodestephen committed:
[[ cpptest ]] Disable memory allocation failure tests on emscripten

livecodestephen committed:
[[ cpptest ]] Correct failure to catch invalid test wrapper

livecodestephen committed:
[[ cpptest ]] Allow emscripten tests to write to a log file

livecodestephen committed:
[[ cpptest ]] Use NODE_JS environment variable

livecodestephen committed:
[[ cpptest ]] Do not build and run the tests as part of the default build target

livecodestephen committed:
[[ cpptest ]] Make Travis run the tests

livecodestephen committed:
[[ cpptest ]] Add `check` targets for Mac and Windows

livecodestephen committed:
[[ cpptest ]] Expand lextable test to cover all the tables

livecodestephen committed:
[[ cpptest ]] Fix compiler warnings

livecodestephen committed:
[[ cpptest ]] Fix some minor memory leaks

These leaks were detected by running test-libFoundation with valgrind.

livecodestephen committed:
[[ cpptest ]] Pull latest googletest

peter-b commented on Makefile:
```diff
@@ -65,6 +65,7 @@ guess_platform_script := \
 
 all: all-$(guess_platform)
+check: check-$(guess_platform)
 
```
It might be worth adding:

``` Makefile
$(MAKE) -C tests/lcb
$(MAKE) -C tests/lcs
```

as commands in the `check` rule, just so that `make check` becomes an easy way to run all the tests.

livecodestephen replied:
Good idea.  Would it be better to add these to the check-% rules so that they can be run for a specific platform?

peter-b replied:
At the moment, those two testsuites do their own (evil) platform detection -- but I don't think it's fair to expect you to fix that in this PR. ;-)

livecodestephen replied:
I was think that that would allow you to run `make check-linux-x86_64` and `make check-linux-x86` and get all the tests both times.

peter-b replied:
Hmm, that sounds like a pretty sensible idea to me.  Please feel free to tweak stuff as appropriate. :+1:

peter-b commented:
@livecodestephen Where do the test logs end up?

livecodestephen commented:
@peter-b: The test logs end up in the build directory, which is also the cwd when they are run.  E.g build-linux-x86_64/livecode/out/Debug/test-libFoundation.log

peter-b commented:
@livecodestephen Okay. Another thing I noticed is that the C++ test output is pretty verbose and doesn't look like the output of our other TAP-based test suites.  Not a blocker from my point of view but if it would be feasible to address it, then that would be nice.  Interestingly, the other test suites use the same log files etc. no matter what target you're building for -- that's probably a bug, since as you rightly point out, it prevents running both the x86 and x86-64 tests on Linux from the same git checkout...

livecodestephen committed:
[[ cpptest ]] Add LCS and LCB tests to the `check` make target

livecodestephen commented:
@peter-b: I'll have a look to see if there's a verbosity setting on the test runner.  You might also want to look at `libcpptest/src/MCTapListener.cpp` as that is where it formats the TAP log files.

peter-b commented:
@livecodestephen I looked over the TAP generation stuff earlier and it all looks fine to me :+1:

livecodestephen commented:
@peter-b It looks like the only way to reduce the output would be to remove the default "listener" and replace it, or redirect the TAP one to stdout.  To remove the default listener you'd need to add something like the following to `gtest_main.cpp`

```
delete listeners.Release(listeners.default_result_printer());
```

peter-b commented:
@livecodestephen Never mind then.

livecodestephen committed:
[[ cpptest ]] Stop travis from explicitly running the LCB and LCS tests

These test are now run as part of the `check-linux` target so they don't need
to be explicitly run by travis.

runrevmark commented:
All the engine changes look good to me... If @livecodefraser could just cast his eye over the gyp, then I think this should be good to go.

livecodefraser commented:
@livecode-vulcan review okay 99f71f6

livecode-vulcan commented:
💙 review by @livecodefraser ok 99f71f6545e292b847a062739653ca8d7ced9048

livecode-vulcan commented:
💥 merge conflict 99f71f6545e292b847a062739653ca8d7ced9048 → aeb98c1ee292f4b89917279c58a921f4d24dfe71

peter-b commented:
@livecode-vulcan review ok 83ebedc

livecodestephen commented:
Somehow I messed up the merge, so have forced pushed back to the commit before the merge.  I'll try merging again.

livecodestephen committed:
Merge branch 'develop' of github.com:livecode/livecode into cpptest

Conflicts:
	.gitmodules

livecodefraser commented:
@livecode-vulcan review okay abe2e6c

livecode-vulcan commented:
💙 review by @livecodefraser ok abe2e6cd8ee4795bcefc5309dbc752f4f407077c

livecode-vulcan commented:
😎 test success abe2e6cd8ee4795bcefc5309dbc752f4f407077c
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/812)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/826)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/813)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/816)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/826)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/836)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/829)
